`Metadata` adjustments and more Algorithms (#27)
wnfs-wg/spec: spec/private-wnfs.md

> 

Reviewer: Recommendation: adding a legend for `CK`, `SK`, and the dashed line would be helpful here.

Reviewer: > - The "decrypted" layer that defines the type of data you can decrypt given the correct keys. Links between blocks in this layer are references in the HAMT data structure at the "encrypted" layer.
> - The "encrypted" layer that defines how all of the encrypted data blocks are organized as IPLD data. Links in this layer are CID-links.

These are often called the "data" and "file" semantic levels. I agree that it's important to clarify that these are encrypted/decrypted in this context, but referencing these layers (or reversed: using these terms and clarifying that files are encrypted) may be a useful since they're existing terms of art. Many file systems will also talk about a "physical" layer, but we're not laying data out on disk.

Author: Hmm. Okay, back when I read the whitepaper, I found the "data" vs. "file" layer distinction confusing.

I think there's already a "data" layer, in, let's say, the public filesystem already, that we're not talking about. That is the underlying IPLD that a file system gets serialized to, which doesn't quite map to file systems themselves. The file layer abstracts over the UnixFS directory shard and the UnixFS file chunking.

I guess in the whitepaper in a way there already existed multiple "data" layers :sweat_smile: 

I'd be in favor of reserving the "data" keyword for another layer I may want to describe in the future. Does that sound reasonable?

Reviewer: > I'd be in favor of reserving the "data" keyword for another layer I may want to describe in the future, specifically all the sharding & chunking. Does that sound reasonable?

Ah, indeed: you want to be able to describe the decrypted data layout as a separate later from the stored encrypted bytes and IPLD layout. (e.g. We probably don't need as much sharding and chunking because it's just big CBOR maps rather than IPLD.) Okay yeah — having separate names for these is a good idea 💯 You're going to laugh that I'm about to suggest YET ANOTHER stack diagram 😛 Something like this (or however you're thinking about it) may be helpful:

```
┌────────────────┬──────────────────┐
│                │                  │
│                │       File       │
│                │                  │
│   Decrypted    ├──────────────────┤
│                │                  │
│                │       Data       │
│                │                  │
├────────────────┼──────────────────┤
│                │                  │
│   Encrypted    │       Data       │
│                │                  │
└────────────────┴──────────────────┘
```

Which implies an "encrypted data" and "decrypted data" layers, plus an implied "decryption" to the file layer...

Wait LOL this is why this kind of exercise is useful. I'm going leave the above text in place even. It needs more cleanup than below, but there's almost more like separate dimensions:

| | Ciphertext | Cleartext |
|--|-------|-------------|
| Data | Block Layout | CBOR Schema |
| File | Namefilters & multivalues | Files & dirs |

Author: I like the stack diagram :) Keeping that idea for once I introduce the "data layer" concept.

## The Encrypted Layer

> Unlike in the public partition, there is no "root" in the private partition. The private file system can contain multiple unrelated private trees at the same time. Thus, we refer to it as a `PrivateForest`. 

Reviewer: We probably want to be very clear here. "Layer" and "root" can have several meanings. There _is_ a root at the data layer, and at least one root at the file layer.

We can go pretty deep in clarifying these layers, but in interest of time for now I'd perhaps clarify something like:

```suggestion
The private file system's encrypted "root" is not the root of the decrypted file system.

The private file system can contain multiple unrelated private trees at the same time. Thus, we refer to it as a `PrivateForest`.
```

BTW, I would like to get a Dark Forest paper on arXiv ahead of StrangeLoop. We'll see if that's even a realistic timeline, but I would love your input or to coauthor such a paper.

Author: Oh yeah, I see how that can be confusing!
Great suggestions, I'm grabbing that :3

> At the encrypted layer, the private forest is a bunch of private data blocks encrypted with different keys. These blocks are supposed to be of size smaller than 256 kilobytes in order to comply with default block size restrictions from IPFS. However, keeping block size small is also useful for reducing metadata leakage - it's less obvious what the file size distribution in the private file system is like, if these files are split into blocks.

Reviewer: No changes required for now; this text is fine as is for our implementation. However, I would suggest reading [RFC2119](https://www.rfc-editor.org/rfc/rfc2119) (it's short!), and using that language. It does change the way you construct sentences, but does help ideas read very clearly and with a clear level of how important a certain detail is.

For example:

> These blocks are SHOULD be less than 256 kilobytes in order to comply with default block size restrictions from IPFS.

or 

> Keeping blocks under 256kb is RECOMMENDED in order to comply with the default block size in IPFS

Author: I want to do a spec-pass over the whole spec with that in mind in the future. I just read the RFC, it's really clear, and MUST/REQUIRED vs. SHOULD/RECOMMENDED vs. MAY makes a lot of sense to me. Thanks!

## Algorithms

All algorithms assume to have access to a `PrivateForest` in their context.

### Namefilter Hash Resolving

`: Hash<Namefilter> -> (Namefilter, Encrypted<PrivateNodeHeader>, Array<Encrypted<PrivateNode>>)`

The private file system is a pointer machine, where pointers are hashes of namefilters.

To resolve a namefilter hash, look up the hash in the HAMT. The resulting key-value pair will give you the full namefilter as well as the private node header and a list of at least one private directory.

Looking up a namefilter hash in the HAMT works by splitting a hash into its nibbles. For example: a hash `0xf199a877d0...` gets split into the nibbles `0xf`, `0x1`, `0x9`, etc.

The nibble order is first taking the 4 most significant bits of and *then* the 4 least significant bits for each hash byte from byte index 0 to 31. This way this matches the usual hex encoding of byte-strings and reading the hex digits off one-by-one.

Each nibble is used as an identifier for a `Node`s child `Node`. Starting at the root, find the root `Node`s child by taking the nibble and computing the index of the child node as

> $$index = popcount(bitmask \land ((1 \ll nibble) - 1))$$

Reviewer: Fine as-is, but might I suggest using a different font for function names:

```suggestion
$$\textsf{index} = \textsf{popcount}(bitmask \land ((1 \ll nibble) - 1))$$
```
<img width="435" alt="Screen Shot 2022-08-10 at 15 29 45" src="https://user-images.githubusercontent.com/1052016/184031652-7283e9cf-87ac-4421-8647-a6bf8cd03cf7.png">

Author: :man_cook: :ok_hand:

If the child is a Node, repeat the process of with the next nibble.

If the child is a HAMT bucket of values, iterate that bucket to find one that has a namefilter that matches the hash of the namefilter. The associated values then contains the ciphertexts and the algorithm is done.

### Private Versioning

> `: (Namefilter, RevisionKey) -> Namefilter`

Reviewer: Interesting; is the `:` with nothing to the left a Rust convention?

Author: No, It's my own convention :sweat_smile: 

Initially it was
```md
### Private Versioning `: (Namefilter, RevisionKey) -> Namefilter`
```

But that looked horrible when rendered, and it produced really unpredictable anchor ids. So I just pushed it to the next line.

I should clarify that in `notation.md`. Do you like/dislike it? Does it help?

Given a `PrivateNodeHeader` it is possible to construct namefilters for newer versions of this private file or directory by stepping the ratchet forward as far as you want to look ahead.

Then, the new namefilter is:

$$saturate(add(deriveKey(inc^n(ratchet)), bareName))$$

Where
- $add$ refers to the [namefilter `add` operation](/spec/namefilter.md#Operation-add)
- $saturate$ refers to the [namefilter `saturate` operation](/spec/namefilter.md#Operation-saturate)
- $deriveKey$ refers to the [skip ratchet `deriveKey` operation](/spec/skip-ratchet.md#Key-Derivation) and
- $inc$ refers to the [skip ratchet increase operation](/spec/skip-ratchet.md#Increasing)

It is possible to choose $n$ in $inc^n(ratchet)$ and due to the properties of the skip ratchet skip ahead versions instead of having to skip one-by-one. When looking for the most recent version of a file, we recommend first skipping to the next small epoch, then the next medium, then large epochs, as long as these revisions exit, and then backtracking once an unpopulated revision is found.

### Path Resolving

`: (PrivateDirectory, Array<string>) -> Hash<Namefilter>`

> Resolving paths in the private file system is always relative. The private partition is actually a forest of trees, and you may not know what other cryptrees exist in there next to what an app may consider the "root". We sometimes refer to the private forest as a "dark forest".

Reviewer: Clarify "relative to what?"

Reviewer: > We sometimes refer to the private forest as a "dark forest".

Let's just make this official somewhere near the top of the document  — it's too good to let pass 😁

Author: I added the 'dark forest' reference to where I'm introducing the 'forest' property itself, below the decrypted layer heading.
